transport/httptransport/http_transport.go

```diff
@@ -0,0 +1,235 @@
+package httptransport
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+
+	"github.com/google/uuid"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/boost/transport/types"
+
+	"github.com/filecoin-project/boost/transport"
+)
+
+var log = logging.Logger("http-transport")
+
+const (
+	// 1 Mb
+	readBufferSize = 1048576
+)
+
+var _ transport.Transport = (*httpTransport)(nil)
+
+func New() *httpTransport {
+	return &httpTransport{}
+}
+
+type httpTransport struct {
+	// TODO Construct and inject an http client here ?
+}
+
+func (h *httpTransport) Execute(ctx context.Context, transportInfo []byte, dealInfo *types.TransportDealInfo) (th transport.Handler, err error) {
+	// de-serialize transport opaque token
+	tInfo := &types.HttpRequest{}
+	if err := json.Unmarshal(transportInfo, tInfo); err != nil {
+		return nil, fmt.Errorf("failed to de-serialize transport info bytes: %w", err)
+	}
+	// check that the outputFile exists
+	fi, err := os.Stat(dealInfo.OutputFile)
+	if err != nil {
+		return nil, fmt.Errorf("output file state error: %w", err)
+	}
+	// validate req
+	if _, err := url.Parse(tInfo.URL); err != nil {
+		return nil, fmt.Errorf("failed to parse url: %w", err)
+	}
+	// do we have more bytes than required already ?
+	fileSize := fi.Size()
+	if fileSize > dealInfo.DealSize {
+		return nil, fmt.Errorf("deal size=%d but file size=%d", dealInfo.DealSize, fileSize)
+	}
+
+	// construct the transfer instance that will act has the transfer handler
+	tctx, cancel := context.WithCancel(ctx)
+	t := &transfer{
+		cancel:         cancel,
+		tInfo:          tInfo,
+		dealInfo:       dealInfo,
+		eventCh:        make(chan types.TransportEvent, 256),
```

Comment:
I think the eventCh should either be unbuffered, or we should have a mechanism that only keeps the latest event (over-writes earlier events)

Reply:
I disagree. The client should get some leeway in terms of consuming it without blocking the transfer.  I don't think over-writes make sense at this layer as these events are simply append only logs of what happened with the transfer. You can do it at the UX layer if you want.

Reply:
In this case the process that reads from the channel is the provider, which reads in a for loop that just passes the event to pubsub. The process writing to the channel is a loop which reads from i/o. So I would expect the process writing to the channel to be much slower. In any case even if the process writing to the channel does get blocked momentarily, I don't believe it matters because the i/o it reads from is being buffered by the OS.

Reply:
I don't have strong opinions here in the absence of having run this in the wild. Happy to make it unbuffered and see how it goes in practise.

Reply:
Ok sounds good 👍

Reply:
Actually, the consumer i.e. Boost SM here reads these transfer events and then pushes them further upwards after transforming there and could be running into back-pressure there or maybe the transformation takes time. I don't want that to block transfers. Remember, the transfer is running a single go-routine and so the transfer will get blocked if it gets blocked writing to the channel here.  We don't need strict synchronisation between the transport and Boot SM here. These notifications are just the append-only log of what's happening with the transfer, NOT used for synchronisation between the transport and Boost SM. Can look into making this a non-blocking queue if you feel strongly against using a buffered channel here.

Reply:
I'd like to see how this works in practise before we decide to change this.

Reply:
Agreed to punt this discussion as Boost evolves.